### ISSUE TITLE
Is there a forum or mailing list?

### ISSUE THREAD

**Issue by Outstep** [Bareflank/hypervisor#871]
Hello All,

I have been learning a lot about various types of Hypervisors ranging from Type-1 (Hyper-V, Xen "XCP-ng, SmartOS, Xvisor, etc.), Type-2 (VirualBox, Byhve, QEMU, Bochs, etc.), Separation Kernels (Muen SK, LynxSecure "commercial SK/Hypervisor Hybrid", seL4, Quest-V, etc..) and now a whole new class of hypervisors like Bareflank, MoRE, kHypervisor, SimpleVisor, HyperPlatform, BitVisor, etc.) that seem to be centered at the firmware level (EFU/UEFI) and single instance hypervisors.

It's a ll pretty exciting stuff.

With that in mind, I am lately interested in something along the lines of LynxSecure but for the open source arena in that it is a Separation Kernel Virtualizing Hypervisor Hybrid.  This is to say that it can act like a SK in that it can partition off resources within a physical computer system into multiple "computers" with each having control of some of the resources and thus OS's can be loaded to run at native speeds with all of the core features like PCI Pass-through, access to come of the memory and some of the CPU's as well as other resources.  In a sense, these types of hypervisors could be classified as Type-0 Hypervisors, I guess.

In any case, I came across BareFlank which seems to run at the EFI level and the video that I saw on it showed that it could boot normally installed OS's like Windows 7, 10, etc. and Linux flavors in 64-bit mode without "paravirtualization" augmentations to the OS drivers. This is very interesting to me.

What I am wondering is if BareFlank has a forum or mailing list so that I could discuss what I have in mind for a project and to see if BareFlank would make a good base from which to begin?

The basic idea, in a first step, is to see if BareFlank could be made to support multiple running OS's running natively with just the hardware resources (no virtualization perhaps yet) and run more like a Separation Kernel.  I would like to see if it is possible to run 3 or 4 OS's concurrently if that is possible, of course with some code modifications and extensions that will need to be developed since BareFlank currently only runs one OS at a time.

Then, maybe next if that went well then things could be extended to be a virtualized by perhaps employing Bhyve ad the VMM for each virtual OS. But this is later. Now I am interested in just a small footprint EFI Separation Kernel Hypervisor.

Anyway, this would be a very nice initial step. Any thoughts, or suggestions would be greatly appreciated.

Cheers and have a good weekend,
Lonnie

**Comment by JaredWright** [comment]
Hi Lonnie,

Thanks for your interest in the project! Bareflank would certainly support your goals of trying to implement a separation kernel that can be launched from EFI.

> What I am wondering is if BareFlank has a forum or mailing list so that I could discuss what I have in mind for a project and to see if BareFlank would make a good base from which to begin?

We use Gitter the primary way to discuss Bareflank related projects, I'll send you an invite to the primary chat room here shortly.

> I would like to see if it is possible to run 3 or 4 OS's concurrently if that is possible, of course with some code modifications and extensions that will need to be developed since BareFlank currently only runs one OS at a time.

You are on the right track here. Bareflank (the repository named "hypervisor") aims to provide an SDK which you can leverage to create more specific hypervisor implementations. It has a built-in extension model that allows you to create your own hypervisors. The [boxy](https://github.com/Bareflank/boxy) project is one such extension that might help get started, because the goal of that project is to implement a hypervisor capable of supporting multiple guest operating systems. Boxy is in proof-of-concept status right now, but it would provide a good starting point for how to use Bareflank to accomplish your goals.

**Comment by Outstep** [comment]
Hi Jared,

While still trying to compile BareFlank from scratch for efi on my Ubuntu
18.04 (x86_64) system, I wanted to respond to your message on my inquiry.
I am digging into Separation Kernels more and still need to determine if
BareFlank laready has the basic functionality or do I need to try to bring
in and integrate an SK from a project like "Muen SK" (
https://muen.codelabs.ch/) or perhaps "NOVA" (http:hypervisor.org) to
handle the context switching stuff.

For the project that I am working on to be a Separation Kernel Virtualizing
Hypervisor Hybrid, I had initially been playing wiht XCP-ng (Xen) and
SmartOS since what I really initially wanted was a very small footprint
hypervisor that could run Linux, Windows, and MAC OS's in minimal
configurations for some applications. Now, the more that I have been
thinking about things, the more that I really like the idea of a Separation
Kernel perhaps like Muen SK, or NOVA, possibly integrated with BareFlank
but keeping things a small as possible.

I think that before I can try to integrate stuff like this, that I probably
need to learn what capabilities BareFlank already has in place.

I will look into Boxy as well to see if I can also build it as a test case.
Cheers,
Lonnie

**Comment by JaredWright** [comment]
> I am digging into Separation Kernels more and still need to determine if
BareFlank laready has the basic functionality...

Bareflank alone does not implement a separation kernel, but rather provides you with everything you would need to implement your own separation kernel from scratch.

> ... or do I need to try to bring
in and integrate an SK from a project like "Muen SK" (
https://muen.codelabs.ch/) or perhaps "NOVA" (http:hypervisor.org) to
handle the context switching stuff

Trying to integrate Bareflank together with an existing separation kernel project will likely lead to issues, because you would end up with two pieces of software competing to manage the same virtualization resources (VMCS, EPT, etc).

> I think that before I can try to integrate stuff like this, that I probably
need to learn what capabilities BareFlank already has in place

To summarize, the base Bareflank repository provides the following capabilities:

* A CMake-based build system and compiler toolchain that lets you compile a VMM
* A collection of runtime support libraries that enable the use of C++
* A set of APIs that let you manage hardware virtualization features (Intel VT-x and VT-d)
* A bare-minimum VT-x configuration that supports turning hardware virtualization features on (i.e. execute the VMXON instruction)
* A loader that can launch the VMM. Bareflank supports loading at boot time from UEFI, or loading after an OS is already running using a driver

**Comment by rianquinn** [comment]
Bareflank can support your goals for a separation kernel today. From UEFI, you can execute the VMM on each core, and then run whatever code you like with some combination of UEFI applications and an extension to the VMM itself. 

Note that the "code" you run on each core is the tricky part. If you attempt to run more than one OS, you would have to ensure the OS's could co-exist. Kernels like Linux, for example, require two timers to work properly, one of the timers being a global timer. Attempting to get multiple OS's running would have to deal with this, either by making mods to the OS to play nice, or to emulate hardware so that the OS's think they are talking to global devices when they are not. If you go down that path however, there really isn't much of a difference between that design, and just using Xen to pin vCPUs to physical cores. 

So just make sure you think about what it is you are trying to accomplish, and what it is about the other projects that do not already achieve the goals you are looking for. These examples exist (we have our own), but you should attempt to determine exactly what it is your attempting to do.

**Comment by rianquinn** [comment]
Going to close this. If you have more questions, please feel free to continue to comment.
